Add temporary path for broadcasting specific shapes to SDPA

mikaylagawarecki opened the pull request:
Adding a hacky path in nested namespace for this specific case for now as discussed with @drisspg, done mostly by copy paste as we don't want to break anything the actual SDPA

```
query (dense) [1, num_head, 1, hidden_dim]
key [B, num_head, *, hidden_dim]
value [B, 1, *, embedding_dim]
output [B, num_head, 1, embedding_dim]
```

### **This should/will most definitely be removed in the near future** if/when we have a more general broadcasting solution for sdpa.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #95346
* #95330

mikaylagawarecki committed:
Add temporary path for broadcasting specific shapes to SDPA

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add temporary path for broadcasting specific shapes to SDPA"




[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/native/nested/cuda/NestedTensorTransformerFunctions.cpp:
```diff
@@ -496,6 +496,109 @@ std::tuple<Tensor, Tensor> _scaled_dot_product_efficient_attention_nestedtensor_
+    {nnz_v_stride, head_v_stride, head_dim_stride},
+    value_impl->get_storage_offsets()[0]);
+  std::tuple<Tensor, Tensor> attention_and_logsumexp=
+      at::_efficient_attention_forward(
```
It seems like at the core of it is calling _efficient_attention_forward with a 0-dim scalar Tensor for one of the cumulative_sequence_length_q or cumulative_sequence_length_k. Why can't we do this from within the existing native functions instead of creating `_scaled_dot_product_efficient_attention_nestedtensor_hack_cuda`?

mikaylagawarecki replied:
`cu_seq_lens_q` a tensor [0, 1, 2, 3, ... batch_size] here

Other things changed include
1) q is a dense tensor so we need to skip the nested related stuff
2) expand q from `{1, 1, num_heads, head_dim_qk}` to `{1, batch_size, num_heads, head_dim_qk}` 
3) adding checks `check_for_seq_len_0_nested_tensor` -> `check_for_seq_len_1_nested_tensor`
4) broadcasting value along num_head dim
5) need to disable the math path can't be used as a fallback for this case

When I discussed with Driss a week-ish ago I think we landed on doing it this way for now, happy to change if there is a better way of encapsulating the above changes without breaking the existing api

cpuhrsch replied:
At a minimum we should avoid adding new native functions for BC breaking reasons. It'll be more difficult to remove. Checking whether value has leading size 1 is something we could branch on within the regular scaled_dot_product_attention native functions.

mikaylagawarecki replied:
Would casing on `q.size(0) == 1 && v.size(1) == 1` in order to go into this path potentially break cases in the real  sdpa where this happens

do we have a new policy on adding private native functions? I know the initial private `_scaled_dot_product_attention` can't be removed now but I don't understand why it would be problematic if this is not used in production

mikaylagawarecki replied:
Hm yea totally agree but I think the tradeoff is that we wouldn't be able to isolate this from the super important and highly used SDPA with my bad change if there's no native_function registration

cpuhrsch commented:
Letting @drisspg take this one after further discussion.

mikaylagawarecki committed:
Update on "Add temporary path for broadcasting specific shapes to SDPA"

Adding a hacky path in nested namespace for this specific case for now as discussed with drisspg, done mostly by copy paste as we don't want to break anything the actual SDPA

```
query (dense) [1, num_head, 1, hidden_dim]
key [B, num_head, *, hidden_dim]
value [B, 1, *, embedding_dim]
output [B, num_head, 1, embedding_dim]
```

### **This should/will most definitely be removed in the near future** if/when we have a more general broadcasting solution for sdpa.




[ghstack-poisoned]

cpuhrsch reviewed (changes requested):
I'm setting this to "request changes" again. Let's talk through it more asynchronously. I think having a divergent API via `_broadcasting_scaled_dot_product_attention` might end up hurting us more than we think.

mikaylagawarecki commented:
@cpuhrsch Ok, I think it is possible to make this a general broadcasting strategy on the nested path for mem_efficient attention for `num_heads` and `batch_size` of q, k, v.

1. still require q, k, v are 4D, have some logic at the start that checks what the real num_head and batch_size should by using 
`[num_heads, batch_size] = at::broadcast_shapes([q.size(0), q.size(1)], [k.size(0), k.size(1)], [v.size(0), v.size(1)])`
2. when`batch_size` is 1
(i) for q: we bypass the `cumulative_and_max_seq_len` call and need to fudge the cu_seq_lens_q with `arange` multiplied by q.size(2) (the size of the sole seq_len) + fudge the Nnz_q with q.size(2) * batch_size
(ii) for k,v: since k and v must have same seq_lens if either one has batch_dim = 1, this indicates that the other one must necessarily have a consistent `seq_len` dim
3) for broadcasting of `num_heads` we need to a fudge the  head_{q/k/v}_stride=0 in the x_storage_as_tensor.as_strided but otherwise I think this isn't problematic

A problem I need to resolve here is q [`1, NH,  1, E_q]` would then need to be nested, so I need a zero copy way of converting a dense tensor to a nested tensor and I don't think this exists, so
4) I imagine this is implementable as a view function(?)

Not sure whether I'm missing anything here @drisspg, does this sound reasonable?

mikaylagawarecki commented:
You are right, it will dispatch to the nested kernel (as is happening here).

mikaylagawarecki commented:
Closing in favor of #95657
